packages/core/content-manager/server/controllers/relations.js

```
@@ -1,64 +1,96 @@
+    const sourceModelUid = component || model;
 
-    if (!modelDef) {
-      return ctx.notFound('model.notFound');
+    const sourceModel = strapi.getModel(sourceModelUid);
+    if (!sourceModel) {
+      return ctx.badRequest("The model doesn't exist");
     }
 
-    const attribute = modelDef.attributes[targetField];
+    const attribute = sourceModel.attributes[targetField];
     if (!attribute || attribute.type !== 'relation') {
-      return ctx.badRequest('targetField.invalid');
+      return ctx.badRequest("This relational field doesn't exist");
     }
 
-    const target = strapi.getModel(attribute.target);
+    const targetedModel = strapi.getModel(attribute.target);
+
+    const offset = Math.max(page - 1, 0) * pageSize;
+    const limit = Number(pageSize);
+
+    const modelConfig = component
+      ? await getService('components').findConfiguration(sourceModel)
+      : await getService('content-types').findConfiguration(sourceModel);
+    const mainField = prop(`metadatas.${targetField}.edit.mainField`, modelConfig) || 'id';
+
+    const query = strapi.db.queryBuilder(targetedModel.uid);
 
-    if (!target) {
-      return ctx.notFound('target.notFound');
+    if (!isNil(q)) {
+      query.search(q);
     }
 
-    if (idsToOmit && Array.isArray(idsToOmit)) {
-      query.filters = {
-        $and: [
-          {
-            id: {
-              $notIn: idsToOmit,
-            },
-          },
-        ].concat(query.filters || []),
-      };
+    if (!isNil(ctx.request.query.filters)) {
+      query.where(convertFiltersQueryParams(ctx.request.query.filters, targetedModel));
     }
 
-    const entityManager = getService('entity-manager');
+    if (!isEmpty(idsToOmit)) {
+      query.where({ id: { $notIn: idsToOmit } });
+    }
 
-    const entities = await entityManager.find(query, target.uid, []);
+    if (entityId) {
+      const joinTable = strapi.db.metadata.get(sourceModelUid).attributes[targetField].joinTable;
```

that code isn't really safe as it won't handle different relation types and join types that the DB supports. We should rather have a utility function to fetch those in the entityManager so we can implement all the necessary usecases of relations

---

What relations do you have in mind that would not work?

---

polymorphic ones or example and realtions not using joinTable (it is possible even if not by default) that's why I think the DB layer should handle so we can hide internal implementation as much as possible

---

yes the route should not be used for polymorphic relations but it should not be the case as the frontend doesn't allow it. For the relations that don't have a join table, indeed then. Would you be available to see how it can be implemented?

---

Yes let's address it too while we do our pair prog :)